player/screenshot.c

```diff
@@ -338,45 +346,64 @@ static struct mp_image *screenshot_get(struct MPContext *mpctx, int mode,
     const struct image_writer_opts *imgopts = mpctx->opts->screenshot_image_opts;
     if (mode == MODE_SUBTITLES && osd_get_render_subs_in_filter(mpctx->osd))
         mode = 0;
-    bool need_add_subs = mode == MODE_SUBTITLES;
 
     if (!mpctx->video_out || !mpctx->video_out->config_ok)
         return NULL;
 
     vo_wait_frame(mpctx->video_out); // important for each-frame mode
 
+    bool use_sw = mpctx->opts->screenshot_sw;
+    bool window = mode == MODE_FULL_WINDOW;
     struct voctrl_screenshot ctrl = {
-        .scaled = mode == MODE_FULL_WINDOW,
+        .scaled = window,
         .subs = mode != 0,
-        .osd = mode == MODE_FULL_WINDOW,
+        .osd = window,
         .high_bit_depth = high_depth && imgopts->high_bit_depth,
         .native_csp = image_writer_flexible_csp(imgopts),
     };
-    if (!mpctx->opts->screenshot_sw)
+    if (!use_sw)
         vo_control(mpctx->video_out, VOCTRL_SCREENSHOT, &ctrl);
     image = ctrl.res;
-    if (image)
-        need_add_subs = false;
 
-    if (!image && mode != MODE_FULL_WINDOW)
-        image = vo_get_current_frame(mpctx->video_out);
-    if (!image) {
+    if (!image && window)
         vo_control(mpctx->video_out, VOCTRL_SCREENSHOT_WIN, &image);
-        mode = MODE_FULL_WINDOW;
+
+    if (!image) {
+        use_sw = true;
+        MP_VERBOSE(mpctx->screenshot_ctx, "Falling back to software screenshot.\n");
     }
+
     if (!image)
-        return NULL;
+        image = vo_get_current_frame(mpctx->video_out);
 
-    if (image->fmt.flags & MP_IMGFLAG_HWACCEL) {
+    /* If hardware decoding is used, we need to download the image from hardware
+     * instead. */
```

Comment:
this isn't 100% accurate because if the vo implements VOCTRL_SCREENSHOT "downloading" it from the gpu 1. still happens 2. but in the VO's code
my suggestion:
```suggestion
    /* vo_get_current_frame() can return a hardware frame, which we have to download first. */
```